src/python/pants/util/memo.py

```diff
@@ -24,6 +26,44 @@ def equal_args(*args, **kwargs):
   return key
 
 
+def coercing_arg_normalizer(type_coercions, base_normalizer, args, kwargs):
+  """Generate a tuple based off of arguments, applying any specified coercions.
+
+  :param dict type_coercions: Map of type -> 1-arg function. If an argument's type matches any
+                              element, the function is called on the argument and stored in the
+                              returned tuple in its place.
+  :param func base_normalizer: A key factory like `equal_args`.
+
+  :rtype: tuple
+  """
+  args_key = base_normalizer(*args, **kwargs)
```

Comment:
There is no contract whatsoever on what the cache key that is output is shaped like. It could be a cryptographic-grade hash value. Normalization should really apply to *args and **kwargs and the transformed versions of those be passed to the `base_normalizer`. As things stand you're presenting a pretend api that actually only works with equal_args et. al. One option is to not pretend this is a generic api. You could take a bool flag instead of base_normalizer, say `per_instance`. If True - pick `per_instance` here and otherwise use `equal_args`. By reparamterizing in that way you could leave the code here as you have it and couple to the knowledge of the output shape of those key factories.